src/ui/EditorNS/jsproxy.cpp

```diff
@@ -0,0 +1,115 @@
+#include "include/EditorNS/jsproxy.h"
+#include <QDebug>
+
+namespace EditorNS {
+
+JsToCppProxy::JsToCppProxy(QObject* parent) : QObject(parent)
+{
+    m_tempQueue = new QQueue<QPair<QString, QVariant>>();
+}
+
+bool JsToCppProxy::hasKey(const QString& key)
+{
+    if(!m_values.contains(key) || m_values[key].isNull()) {
+        return false;
+    }
+    return true;
+}
+
+void JsToCppProxy::pushQueuedMessages()
+{
+    while(!m_tempQueue->isEmpty()) {
+        QPair<QString, QVariant> msg = m_tempQueue->dequeue();
+        emit sendMsgInternal(msg.first, msg.second);
+    }
+    delete m_tempQueue;
+}
+
+void JsToCppProxy::sendMsg(const QString& msg, const QVariant& data)
+{
+    if(!m_ready) {
+        m_tempQueue->enqueue(qMakePair(msg, data));
+    }else {
+        emit sendMsgInternal(msg, data);
+    }
+}
+
+void JsToCppProxy::sendEditorEvent(QString msg, QVariant data)
+{
+    if (msg == "J_EVT_READY" && m_ready == false) {
+        m_ready = true;
+        pushQueuedMessages();
+    }else if(msg == "J_EVT_CLEAN_CHANGED") {
+        m_values["clean"] = data;
+    }else if(msg == "J_EVT_CURRENT_LANGUAGE_CHANGED") {
+        m_values["language"] = data;
+    }
```

Comment:
J_EVT_CLEAN_CHANGED and J_EVT_CURRENT_LANGUAGE_CHANGED don't belong here because we don't want to know about all possible events here. J_EVT_READY was only a special case.

There's still room for improvements but I actually prefer how it was done before... you can revert this commit (fcd8dd022729615d94f7d81eb22748e9f15b19e6)